# Notebook 04 — Hypothesis Testing Analysis

## Research Questions
**Q2 (Inferensi & Hipotesis):** Apakah rata-rata jumlah komentar berbeda secara signifikan antara PR yang *merged* vs *unmerged* di repositori `pandas-dev/pandas`?

## Member
- **Nama:** Bethelina Imanuella Yesaya
- **Role:** Hypothesis Analyst (Member D)

## AI Usage Disclosure

**Member:** Bethelina Imanuella Yesaya — Hypothesis Analyst (Member D)  
**Tools used:** Gemini, ChatGPT

| Task | Tool | Prompt summary | Output modified? |
|---|---|---|---|
| Membuat kode Two-Sample Z-Test | Gemini | "bantu saya membuat kode Two-Sample Z-Test di python" | Ya |
| Menjelaskan konsep statistik | ChatGPT | "jelaskan apa itu p-value, Two-Sample Z-Test, Two-Tailed Test, dll" | Ya |
| Perbaikan kode Python | ChatGPT | "bantu saya merapikan kode python berikut" | Ya |

**Written entirely without AI:** Penjabaran prosedur 6 langkah uji hipotesis, perumusan H0 dan H1.

In [ ]:
import pandas as pd
import numpy as np
from scipy import stats

url_link_raw = "https://raw.githubusercontent.com/UsmanNaf/stat-audit-FinalProject_Group10-sti-2025/refs/heads/main/data/clean/pr_clean.csv"

# Load data
df_pr = pd.read_csv(url_link_raw)
print(f"Data berhasil dimuat. Total baris data: {len(df_pr)}")

# Pisahkan data berdasarkan status PR
comments_merged = df_pr[df_pr['status'] == 'merged']['comments']
comments_unmerged = df_pr[df_pr['status'] == 'unmerged']['comments']

# Statistik deskriptif
x_bar1 = comments_merged.mean()
x_bar2 = comments_unmerged.mean()

n1 = len(comments_merged)
n2 = len(comments_unmerged)

sigma1 = comments_merged.std()
sigma2 = comments_unmerged.std()

# Two-Sample Z-Test
pooled_se = np.sqrt((sigma1**2 / n1) + (sigma2**2 / n2))
z_stat = (x_bar1 - x_bar2) / pooled_se
p_value = 2 * (1 - stats.norm.cdf(abs(z_stat)))

# Output
print("\n" + "=" * 60)
print("HASIL OUTPUT PENGUJIAN TWO-SAMPLE Z-TEST")
print("=" * 60)

print(f"PR Merged")
print(f"  Mean              : {x_bar1:.4f}")
print(f"  Std. Deviation    : {sigma1:.4f}")
print(f"  Sample Size (n1)  : {n1}")

print()

print(f"PR Unmerged")
print(f"  Mean              : {x_bar2:.4f}")
print(f"  Std. Deviation    : {sigma2:.4f}")
print(f"  Sample Size (n2)  : {n2}")

print("-" * 60)
print(f"Z-Statistic         : {z_stat:.4f}")
print(f"P-Value             : {p_value:.4e}")
print(f"Decision            : {'Reject H0' if p_value < 0.05 else 'Fail to Reject H0'}")
print("=" * 60)

Data berhasil dimuat. Total baris data: 789

HASIL OUTPUT PENGUJIAN TWO-SAMPLE Z-TEST
PR Merged
  Mean              : 1.2380
  Std. Deviation    : 1.2572
  Sample Size (n1)  : 500

PR Unmerged
  Mean              : 2.1246
  Std. Deviation    : 3.3195
  Sample Size (n2)  : 289
------------------------------------------------------------
Z-Statistic         : -4.3631
P-Value             : 1.2825e-05
Decision            : Reject H0


## Prosedur 6 Langkah Uji Hipotesis


---


**1. Mengidentifikasi Klaim Awal**

Terdapat perbedaan rata-rata jumlah komentar antara PR *merged* dan PR *unmerged.*


---


**2. Merumuskan Hipotesis**

H0 : μ*merged* = μ*unmerged* (artinya tidak terdapat perbedaan rata-rata jumlah komentar antara PR *merged* dan PR *unmerged*)

H1 : μ*merged* ≠ μ*unmerged* (artinya terdapat perbedaan rata-rata jumlah komentar antara PR *merged* dan PR *unmerged*)


---


**3. Menentukan Tingkat Signifikansi**

Tingkat signifikasi yang digunakan adalah α = 0.05 (5%)

Yang berarti risiko melakukan Galat Tipe I (menolak H0 padahal benar) sebesar 5%.


---


**4. Mengumpulkan Data**

Data diambil dari dataset PR repositori *pandas-dev/pandas* yang telah melalui proses pembersihan data, dengan ringkasan data yang digunakan sebagai berikut:

- **PR Merged**

Jumlah sampel (n) = 500

Rata-rata Komentar = 1.2380

- **PR Unmerged**

Jumlah sampel (n) = 289

Rata-rata Komentar = 2.1246

Variabel yang dianalisis adalah jumlah komentar (*comments*) pada setiap PR.

---


**5. Menghitung Nilai Statistik Uji dan p-value**

Pengujian dilakukan menggunakan Two-Sample Z-Test dengan statistik uji:
$$Z=\frac{\bar{x}_1-\bar{x}_2}{\sqrt{\frac{\sigma_1^2}{n_1}+\frac{\sigma_2^2}{n_2}}}$$

Hasil perhitungannya menunjukkan:

- $\bar{x}_1$ (PR *merged*) = 1.2380
- $\bar{x}_2$ (PR *unmerged*) = 2.1246
- $Z = -4.3631$
- $p$-value $= 1.2825 \times 10^{-5}$

Nilai p-value didapat sangat kecil dibandingkan tingkat signifikansi yang digunakan.

---


**6. Mengambil Keputusan**

Karena:

$$
p\text{-value} = 1.2825 \times 10^{-5} < \alpha = 0.05
$$

maka **H₀ ditolak (Reject H₀)**.

Dengan demikian, rata-rata jumlah komentar pada pull request *merged* dan *unmerged* di repositori *pandas-dev/pandas* berbeda secara signifikan.


---


## Interpretasi Hasil

Hasil pengujian menunjukkan bahwa PR yang berstatus *unmerged* memiliki rata-rata komentar sebesar 2,1246, lebih tinggi dibandingkan PR yang berstatus *merged* yang memiliki rata-rata komentar sebesar 1,2380. Hal tersebut berarti PR yang tidak berhasil di-merge cenderung melibatkan diskusi yang lebih panjang atau lebih banyak *feedback* dari maintainer dan kontributor lain. Beberapa faktor dari banyaknya komentar bisa dari adanya revisi yang belum terselesaikan atau kualitas kode yang perlu perhatian lebih lanjut dari maintainer. Hasil ini menunjukkan bahwa tingginya jumlah komentar dapat menjadi indikator awal adanya hambatan dalam proses review. Oleh karena itu, informasi ini dapat dimanfaatkan maintainer untuk mengidentifikasi PR yang berpotensi mengalami kesulitan dan juga untuk memberikan *feedback* yang lebih terarah.